### DISCUSSION TITLE
How to represnt this lua table? tbl = { ["x"] = "xxx", ["y"] = "yyy", ["z"] = "zzz"}

### DISCUSSION THREAD

**Discussion by andreacfromtheapp** [MunifTanjim/nui.nvim#324] [Q&A] [answered]
Hi all,

I'm trying to use NuiTable to represent the following table but can't manage to:

```lua
local tbl = { 
    ["1"] = "foo",
    ["2"] = "bar",
    ["3"] = "baz"
}

dada_table = NuiTable({
    bufnr = dada_pane.bufnr,
    columns = {
        {
            align = "center",
            header = "DADA",
            columns = {
                { "key", header = "ID" },
                { "value", header = "Name" },
            },
        },
    },
    data = { tbl },
})
```

What would be the code to do it please?

Thank you

**Comment by MunifTanjim** [comment, marked as answer]
First convert your data shape:

```lua
local tbl = {
  ["1"] = "foo",
  ["2"] = "bar",
  ["3"] = "baz",
}

local data = {}

for id, name in pairs(tbl) do
  table.insert(data, { id = id, name = name })
end

--[[
local data = {
  { id = "1", name = "foo" },
  { id = "2", name = "bar" },
  { id = "3", name = "baz" },
}
--]]
```

Then update your table definition:

```lua
dada_table = NuiTable({
  bufnr = dada_pane.bufnr,
  columns = {
    {
      align = "center",
      header = "DADA",
      columns = {
        { accessor_key = "id", header = "ID" },
        { accessor_key = "name", header = "Name" },
      },
    },
  },
  data = data,
})
```

Ref: https://github.com/MunifTanjim/nui.nvim/blob/main/lua/nui/table/README.md

**Comment by andreacfromtheapp** [comment]
Hi @MunifTanjim :)

firstly: thanks for Nui!

secondly: didn't think of that!

lastly, one more question please:

- since the initial table is populated on the fly by a nvim `jobstart`, would NuiTable refresh on new data being inserted? Or would I need to update the layout every time? or?

Please and thank you

**Reply by MunifTanjim** [reply]
You can keep the reference of `data` somewhere and update it when you want to add/remove items.

For example:

```lua
local Popup = require("nui.popup")
local Table = require("nui.table")
local Text = require("nui.text")

local popup = Popup({
  position = "50%",
  size = {
    width = 40,
    height = 20,
  },
  relative = "editor",
})

local data = {
  { firstName = "John", lastName = "Doe", age = 42 },
  { firstName = "Jane", lastName = "Doe", age = 27 },
}

local tbl = Table({
  bufnr = popup.bufnr,
  columns = {
    {
      align = "center",
      header = "Name",
      columns = {
        { accessor_key = "firstName", header = "First" },
        {
          id = "lastName",
          accessor_fn = function(row)
            return row.lastName
          end,
          header = "Last",
        },
      },
    },
    {
      align = "right",
      accessor_key = "age",
      cell = function(cell)
        return Text(tostring(cell.get_value()), "DiagnosticInfo")
      end,
      header = "Age",
    },
  },
  data = data,
})

tbl:render()
popup:mount()

vim.defer_fn(function()
  table.insert(data, { firstName = "Who", lastName = "Doctor", age = 1000 })
  tbl:render()
end, 2000)
```

This will add a new row after 2 seconds.

Similarly you can clear data by doing:

```lua
  for idx in ipairs(data) do
    data[idx] = nil
  end
  tbl:render()
```

**Comment by andreacfromtheapp** [comment]
@MunifTanjim two more questions please :)

this is what I'm trying to achieve but I'm not happy with my design:

<img width="960" alt="Screenshot 2024-01-18 at 12 25 42 PM" src="https://github.com/MunifTanjim/nui.nvim/assets/3269984/169d9731-8df8-4701-bdf1-c983957b43ec">

code:

```lua
local Layout = require("nui.layout")
local NuiSplit = require("nui.split")
local NuiTable = require("nui.table")
-- local NuiText = require("nui.text")
-- local shreds = require("chuck-nvim.core.shreds").table

local M = {}

local shreds = {
    { id = "1", name = "footer.ck" },
    { id = "2", name = "baritalia.ck" },
    { id = "3", name = "bazzing.ck" },
}

M.event = require("nui.utils.autocmd").event

M.shred_pane = NuiSplit({
    ns_id = "shred_pane",
    enter = true,
})

M.chuck_pane = NuiSplit({
    ns_id = "chuck_pane",
    enter = true,
})

M.shreds_table = NuiTable({
    bufnr = M.shred_pane.bufnr,
    columns = {
        {
            align = "center",
            header = "Shreds",
            columns = {
                { align = "center", width = 26, accessor_key = "id", header = "Shred ID" },
                { align = "center", width = 26, accessor_key = "name", header = "Shred Name" },
            },
        },
    },
    data = shreds,
})

M.chuck_layout = Layout(
    {
        position = "right",
        relative = "editor",
        size = "30%",
    },
    Layout.Box({
        Layout.Box(M.chuck_pane, { size = "55%" }),
        Layout.Box(M.shred_pane, { size = "45%" }),
    }, {
        dir = "col",
    })
)

M.update_layout = (
    Layout.Box({
        Layout.Box(M.shred_pane, { size = "45%" }),
        Layout.Box(M.chuck_pane, { size = "55%" }),
    }, {
        dir = "col",
    })
)

return M
```
Questions:

- 1 is there a way to not show line numbers in NuiSplit? (`:set nonumber` and `:set norelativenumber` equivalent?)
- 2 how to achieve a similar list with only one entry per line? it takes too much space rn.

Please and thank you

**Reply by andreacfromtheapp** [reply]
ps: I would not care about headings with NuiLine or NuiText. the table was my first idea but.

**Reply by MunifTanjim** [reply]
> 1 is there a way to not show line numbers in NuiSplit?

Split has a `.winid` associated with it. After mount, you can do this:

```lua
vim.wo[split.winid].number = false

-- or 

vim.api.nvim_set_option_value('number', false, { scope = "local", win = split.winid ))
```

where `split` in you case would be `M.shred_pane` or `M.chuck_pane`.

**Reply by andreacfromtheapp** [reply]
pps: this is how I call it for now but it's fugly:

```lua
function M.chuck_split(cmd, logfile)
    layout.chuck_layout:mount()
    layout.chuck_pane:on(layout.event.BufEnter, function()
        vim.cmd("set nonumber")
        vim.cmd("set norelativenumber")
        start_chuck(cmd, logfile)
    end, { once = true })
    layout.shred_pane:on(layout.event.BufEnter, function()
        vim.cmd("set nonumber")
        vim.cmd("set norelativenumber")
        -- shred_lines(logfile)
    end, { once = true })
    vim.cmd("wincmd w")

    -- FIX: this is a workaound until I can figure out the event to use to
    -- trigger the callback functions on layout mount in the above code.
    vim.cmd("wincmd w")
    vim.cmd("wincmd w")
    vim.cmd("wincmd w")

    layout.chuck_layout:update(layout.update_layout)
    layout.shreds_table:render() -- FIX: nui table idea
end
```

I'd like to get rid of all the vim.cmd workarounds

**Reply by MunifTanjim** [reply]
> 2 how to achieve a similar list with only one entry per line? it takes too much space rn.

You mean table without the borders? Border customization for `NuiTable` is not yet supported.

**Reply by andreacfromtheapp** [reply]
```
    vim.api.nvim_set_option_value("number", false, { scope = "local", win = layout.chuck_pane.winid })
    vim.api.nvim_set_option_value("number", false, { scope = "local", win = layout.shred_pane.winid })
```

this is not working. it removes the line number but it places a 0 on top line and I cannot see my utput. I'll stick with vim.cmd for those at the moment.

**Reply by andreacfromtheapp** [reply]
> > 2 how to achieve a similar list with only one entry per line? it takes too much space rn.
> 
> You mean table without the borders? Border customization for `NuiTable` is not yet supported.

not necessarily a NuiTable. I'd be happy to use a for loop on NuiLine or NuiText to display "1 foo.ck" and so on on their own lines. the table was my first idea, but I can ditch it for better layout.

**Reply by MunifTanjim** [reply]
> this is not working. it removes the line number but it places a 0 on top line and I cannot see my utput. I'll stick with vim.cmd for those at the moment.

Maybe try setting `relativenumber` too?

```lua
    vim.api.nvim_set_option_value("number", false, { scope = "local", win = layout.chuck_pane.winid })
    vim.api.nvim_set_option_value("number", false, { scope = "local", win = layout.shred_pane.winid })
    vim.api.nvim_set_option_value("relativenumber", false, { scope = "local", win = layout.chuck_pane.winid })
    vim.api.nvim_set_option_value("relativenumber", false, { scope = "local", win = layout.shred_pane.winid })

```

**Reply by andreacfromtheapp** [reply]
oops.... it works ofc. please consider making that a split option :) I'd be happy to open a feature request

**Reply by MunifTanjim** [reply]
Actually I forgot that it already has a shortcut for setting it:

https://github.com/MunifTanjim/nui.nvim/blob/main/lua/nui/split/README.md#win_options

```lua
local split = Split({
  win_options = {
    number = false,
    relativenumber = false,
  }
})
```

**Reply by andreacfromtheapp** [reply]
amazing! thanks :)

**Comment by andreacfromtheapp** [comment]
@MunifTanjim not sure you've seen this in all the messages. Reposting:

> > 2 how to achieve a similar list with only one entry per line? it takes too much space rn.
> 
> You mean table without the borders? Border customization for `NuiTable` is not yet supported.

not necessarily a NuiTable. I'd be happy to use a for loop on NuiLine or NuiText to display "1 foo.ck" and so on on their own lines. the table was my first idea, but I can ditch it for better layout.

**Reply by MunifTanjim** [reply]
You can try using `NuiTree`.

```lua
local tree = NuiTree({
  bufnr = bufnr,
  nodes = {
    NuiTree.Node({ text = "a" }),
    NuiTree.Node({ text = "b" }),
  },
})

tree:render()
```

`text` for the node can be `string` or `NuiText` or `NuiLine`. You can do even more complex stuffs using `prepare_node`: https://github.com/MunifTanjim/nui.nvim/tree/main/lua/nui/tree#prepare_node

**Reply by MunifTanjim** [reply]
Or if you want to directly render `NuiText` or `NuiLine` on a buffer, nothing would stop you from doing that. That is also possible.

**Reply by andreacfromtheapp** [reply]
NuiTree does what I need! It can also add/remove/get changes! :) Can't figure out how to pass my table to it tho. :(

```lua
local shreds = {
    { id = "1", name = "footer.ck" },
    { id = "2", name = "baritalia.ck" },
    { id = "3", name = "bazzing.ck" },
}

local function prepare_node(node)
    local line = NuiLine()
    line:append(string.rep("  ", node:get_depth() - 1))
    if node:has_children() then
        line:append(node:is_expanded() and " " or " ")
    else
        line:append("  ")
    end

    line:append(node.text)

    return line
end

M.shreds_tree = NuiTree({
    bufnr = M.shred_pane.bufnr,
    nodes = {
        NuiTree.Node({ text = prepare_node(shreds) }),
    },
})
```

sorry about the many questions, it is my first nvim plugin and first time lua.

**Reply by MunifTanjim** [reply]
```lua
local shreds = {
    { id = "1", name = "footer.ck" },
    { id = "2", name = "baritalia.ck" },
    { id = "3", name = "bazzing.ck" },
}

local nodes = {}

for _, shred in ipairs(shreds) do
  table.insert(nodes, NuiTree.Node(shred))
end

M.shreds_tree = NuiTree({
    bufnr = M.shred_pane.bufnr,
    nodes = nodes,
    prepare_node = function(node)
      return NuiLine({
        NuiText(node.id),
        NuiText("  "),
        NuiText(node.name),
      })
    end,
})
```

**Reply by MunifTanjim** [reply]
If you're completely new in Lua and Neovim plugin development, I'd recommend seeing how other plugins are using `NuiTree`, for example: https://github.com/nvim-neo-tree/neo-tree.nvim

List of plugins using `nui.nvim`: https://github.com/MunifTanjim/nui.nvim/wiki/nui.nvim-in-the-wild

**Reply by MunifTanjim** [reply]
Also, GitHub search: https://github.com/search?q=nui.tree+language%3ALua&type=code&l=Lua

**Reply by andreacfromtheapp** [reply]
Thank you soooo much! Now I need to figure out how to keep it up to date (add/remove specific nodes). I've been lurking other plugins but dind't know neo-tree did use nui! Will definitely check it out :) Thank you again!!!!

**Reply by MunifTanjim** [reply]
> Now I need to figure out how to keep it up to date (add/remove specific nodes).

Check: https://github.com/MunifTanjim/nui.nvim/tree/main/lua/nui/tree#methods

Specifically look for these methods: `add_node`, `remove_node` and `set_nodes`.

**Reply by andreacfromtheapp** [reply]
will definitely look into those as soon as I manage to get the function/job to update the main `shreds_table` working. here's the desired effect (although I'd like to add split pane titles like popups do):

<img width="960" alt="Screenshot 2024-01-18 at 3 48 42 PM" src="https://github.com/MunifTanjim/nui.nvim/assets/3269984/7eaee464-6580-4bc8-a9e2-e98c59914251">

while we are on this discussion: is there a better way to run callback functions than this? Ideally I would run them on mount of each pane instead of moving the focus around and BufEnter. I tried BufNew, BufAdd, and BufCreate but nope.

```lua
function M.chuck_ui(cmd, logfile)
    layout.chuck_layout:mount()

-- from here
    layout.chuck_pane:on(layout.event.BufEnter, function()
        start_chuck(cmd, logfile)
    end, { once = true })

    layout.shred_pane:on(layout.event.BufEnter, function()
        shred_lines(logfile)
    end, { once = true })

    vim.cmd("wincmd w")

    -- FIX: this is a workaound until I can figure out the event to use to
    -- trigger the callback functions on layout mount in the above code.
    vim.cmd("wincmd w")
    vim.cmd("wincmd w")
    vim.cmd("wincmd w")
-- to here

    layout.chuck_layout:update(layout.update_layout)
end
```

**Reply by andreacfromtheapp** [reply]
it seems that after I added win_options and but_options (as read only) it works fine! removed all those ugly wincmd w

**Comment by andreacfromtheapp** [comment]
> > Now I need to figure out how to keep it up to date (add/remove specific nodes).
> 
> Check: https://github.com/MunifTanjim/nui.nvim/tree/main/lua/nui/tree#methods
> 
> Specifically look for these methods: `add_node`, `remove_node` and `set_nodes`.

Hi @MunifTanjim, after all I think I need something repopulating the tree each time the table is updated:  https://github.com/MunifTanjim/nui.nvim/discussions/71#discussioncomment-1840821 but can't figure out how to implement it.

Can I please ask for one more help? Fingers crossed this should be it.

Please and thank you

EDIT:

this is where I'd use it

from https://github.com/gacallea/chuck-nvim/blob/main/lua/chuck-nvim/ui/layout.lua

```lua
[...]
M.set_expanded_nodes = function(tree)
    local nodes = mknodes()
    for _, id in ipairs(nodes or {}) do
        local node = tree:get_node(id)
        if node ~= nil then
            node:expand()
        end
    end
end
[...]
```

from https://github.com/gacallea/chuck-nvim/blob/main/lua/chuck-nvim/core/utils.lua

```lua
local layout = require("chuck-nvim.ui.layout")
local shreds = require("chuck-nvim.core.shreds")

[...]

local function shred_lines(logfile)
    local file = assert(io.open(logfile, "r"))
    file:seek("end")
    local it = file:lines()
    while vim.fn.getchar(0) == 0 do
        local line = it()
        if line then
            shreds.set_table(line)
            layout.shreds_tree.set_expanded_nodes(layout.shreds_tree)
            layout.shreds_tree:render()
        end
    end
    file:close()
end

[...]
```

**Reply by andreacfromtheapp** [reply]
ps: just for completeness sake -- I hope it helps -- here's the plugin and what I'm trying to do https://github.com/gacallea/chuck-nvim 

and https://github.com/gacallea/chuck-nvim/issues/5#issuecomment-1898627699

**Reply by andreacfromtheapp** [reply]
pps: I just realized taht I may have bigger issue: the shreds table isn't updating in that while loop. ~~the loop is valid and it works if I `io.write(line)` to a split buffer but it doesn't work as inteded above yet.~~ arg!

**Reply by MunifTanjim** [reply]
I'm really sorry, but these are way too much information.

I'd recommend you to try out `NuiTree` in a simple script, and play with it. If you have any confusions or questions about how to do a particular thing with `nui.nvim`, I'd be happy to help.

But if you include codes from your plugin, it gets very hard for me to debug/understand it... because the issue can be caused by things that has nothing to do with `nui.nvim`.

---

If you're trying to do something with `NuiTree` and can't figure out how to do it, please include a simple script showing the problem. Then I can point out how to solve it.

**Reply by andreacfromtheapp** [reply]
my apologies about the code dump and too much information. and thank you for your help. you made 100% of the work for this to function as it does, and I'm grateful. As soon as I fix the shreds_table inserting it may just works as I invoke render after each insertion (when it will work). thank you so much again and sorry.

**Reply by MunifTanjim** [reply]
Hey, no worries. No need to be sorry. It's just that if you share minimal example, it's easier for me to help you. I'm happy to help.

**Reply by andreacfromtheapp** [reply]
Hi @MunifTanjim :)

I've finally managed (with help from nvim matrix channel) to make the code work up to the point to populate the table used by NuiTree! Now the issue I have is how to refresh the tree each time this runs:

```lua
local function shred_set(line)
  local action = set_action(line)
  if line and action then
    shreds.set_table(line, action)
    -- here --  something like this?
    layout.shreds_tree.update()
  end
end
```

the initial layout is dealt here (for info's sake):

```lua
function M.chuck_ui(cmd, logfile)
  layout.chuck_layout:mount()
  layout.chuck_pane:on(events.BufEnter, function()
    start_chuck(cmd, logfile)
  end, { once = true })
  layout.shred_pane:on(events.BufEnter, function()
    shred_lines(logfile)
  end, { once = true })
  vim.cmd("wincmd w")

  layout.chuck_layout:update(layout.update_layout)
  layout.shreds_tree:render()
end
```

I know I have to use `set_nodes` in my layout.lua but I can't manage. Is that a new `update()` a function that runs `mknodes()` (this basically read the table and `table.insert(nodes, NuiTree.Node(shred))` again?

Please and thank you

**Reply by andreacfromtheapp** [reply]
this is what I've tried but ehm

`M.shreds_tree` and `mknodes()` are defined in `layout.lua` and used to initialize the tree on `M.shreds_tree:render()`. The logic seems to be there but the code is pseudo-code at best.

```lua
M.shreds_tree_update = M.shreds_tree:set_nodes(mknodes())
```

```lua
local function shred_set(line)
  local action = set_action(line)
  if line and action then
    shreds.set_table(line, action)
    layout.shreds_tree_update:render()
  end
end
```

**Reply by andreacfromtheapp** [reply]
getting closer, I think :)

```lua
local function shred_set(line)
  local action = set_action(line)
  if line and action then
    shreds.set_table(line, action)
    local nodes = layout.mknodes()
    layout.shreds_tree:set_nodes(nodes)
  end
end
```

**Reply by andreacfromtheapp** [reply]
Ok I see the problem now. EACH line is a NuiTree.Node(shred) here.... I should probably iterate the table and replace stuff by Node.id.

Edit: unless there is a way for set_nodes to call prepare_node automatically in 

```lua
M.shreds_tree = NuiTree({
  bufnr = M.shred_pane.bufnr,
  nodes = M.mknodes(),
  prepare_node = function(node)
    return NuiLine({
      NuiText("id: "),
      NuiText(node.id, "DiagnosticOk"),
      NuiText("  "),
      NuiText("name: "),
      NuiText(node.name, "DiagnosticOk"),
      NuiText("  "),
      NuiText("time: "),
      NuiText("01:08", "DiagnosticOk"),
    })
  end,
})
```

**Reply by andreacfromtheapp** [reply]
@MunifTanjim 

or I could skip the table entirely!!!!! now I need to figure out remove and replace.... :D

```lua
local function set_node(line, action)
  if action ~= nil then
    local pattern = ".*(%d+)%s+%((.-)%)"
    local shred_id, shred_name = line:match(pattern)

    if shred_id ~= nil and shred_name ~= nil and shred_name:match(".ck") then
      if action == "add" then
        layout.shreds_tree:add_node(NuiTree.Node({ id = shred_id, name = shred_name }))
      end
      -- if action == "replace" then end
      if action == "remove" then
        layout.shreds_tree:remove_node(shred_id)
      end
    end
  end
end
```

**Reply by andreacfromtheapp** [reply]
almost there!!! I would ask help on the last thing: how do I empty the whole tree please? trying with set_nodes() to empty but nope.


```lua
-- https://neovim.io/doc/user/diagnostic.html#diagnostic-highlights
M.shreds_tree = NuiTree({
  bufnr = M.shred_pane.bufnr,
  nodes = {},
  get_node_id = function(node)
    if node.id then
      return node.id
    end
  end,
  prepare_node = function(node)
    return NuiLine({
      NuiText("id: "),
      NuiText(tostring(node.id), "DiagnosticOk"),
      NuiText("  "),
      NuiText("name: "),
      NuiText(node.name, "DiagnosticOk"),
    })
  end,
})

function M.set_node(line, action)
  if action ~= nil then
    local pattern = ".*(%d+)%s+%((.-)%)"
    local shred_id, shred_name = line:match(pattern)

    if shred_id ~= nil and shred_name ~= nil and shred_name:match(".ck") then
      if action == "add" then
        M.shreds_tree:add_node(NuiTree.Node({ id = shred_id, name = shred_name }))
      end
      if action == "replace" then
        M.shreds_tree:remove_node(shred_id)
        M.shreds_tree:add_node(NuiTree.Node({ id = shred_id, name = shred_name }))
      end
      if action == "remove" then
        M.shreds_tree:remove_node(shred_id)
      end
      if action == "clear" then
        M.shreds_tree:set_nodes(NuiTree.Node({}))
      end
    end
  end
end
```

**Reply by MunifTanjim** [reply]
> how do I empty the whole tree please? trying with set_nodes() to empty but nope.

As the doc says, the first param of `set_nodes` is expected to be a list of nodes.

https://github.com/MunifTanjim/nui.nvim/tree/main/lua/nui/tree#treeset_nodes

```lua
tree:set_nodes(NuiTree.Node({}))
```

This is passing a single node. Which is wrong.

If you want to clear all nodes, you should pass an empty list:

```lua
tree:set_nodes({})
```

**Reply by andreacfromtheapp** [reply]
I did try that at first and it didn't work, because I was doing it in the wrong place.... thank you a million times for your help and patience and my apologies for so many questions.

```lua
function M.set_node(line, action)
  if action ~= nil then
    local pattern = ".*(%d+)%s+%((.-)%)"
    local shred_id, shred_name = line:match(pattern)

    if action == "clear" then
      M.shreds_tree:set_nodes({})
    end

    if shred_id ~= nil and shred_name ~= nil and shred_name:match(".ck") then
      if action == "add" then
        M.shreds_tree:add_node(NuiTree.Node({ id = shred_id, name = shred_name }))
      end
      if action == "replace" then
        M.shreds_tree:remove_node(shred_id)
        M.shreds_tree:add_node(NuiTree.Node({ id = shred_id, name = shred_name }))
      end
      if action == "remove" then
        M.shreds_tree:remove_node(shred_id)
      end
    end
  end
end
```
